rir/src/runtime/TypeFeedback.cpp

```diff
@@ -1,108 +1,97 @@
+        if (invalidateWhenFull) {
             invalid = true;
+            return true;
+        }
     }
+    return false;
 }
 
-SEXP ObservedCallees::getTarget(const Code* code, size_t pos) const {
+SEXP ObservedCallees::getTarget(const Function* function, size_t pos) const {
     assert(pos < numTargets);
-    return code->getExtraPoolEntry(targets[pos]);
+    return function->body()->getExtraPoolEntry(targets[pos]);
 }
 
-FeedbackOrigin::FeedbackOrigin(rir::Code* src, Opcode* p)
-    : offset_((uintptr_t)p - (uintptr_t)src), srcCode_(src) {
-    if (p) {
-        assert(p >= src->code());
-        assert(p < src->endCode());
-        assert(pc() == p);
-    }
+FeedbackOrigin::FeedbackOrigin(rir::Function* function, FeedbackIndex index)
+    : index_(index), function_(function) {
+    assert(function->typeFeedback()->isValid(index));
 }
 
 DeoptReason::DeoptReason(const FeedbackOrigin& origin,
                          DeoptReason::Reason reason)
-    : reason(reason), origin(origin) {
-    switch (reason) {
-    case DeoptReason::Typecheck:
-    case DeoptReason::DeadCall:
-    case DeoptReason::CallTarget:
-    case DeoptReason::ForceAndCall:
-    case DeoptReason::DeadBranchReached: {
-        assert(pc());
-        auto o = *pc();
-        assert(o == Opcode::record_call_ || o == Opcode::record_type_ ||
-               o == Opcode::record_test_);
-        break;
-    }
-    case DeoptReason::Unknown:
-    case DeoptReason::EnvStubMaterialized:
-        break;
-    }
-}
+    : reason(reason), origin(origin) {}
 
 void DeoptReason::record(SEXP val) const {
-    srcCode()->function()->registerDeoptReason(reason);
+    origin.function()->registerDeoptReason(reason);
 
     switch (reason) {
     case DeoptReason::Unknown:
         break;
     case DeoptReason::DeadBranchReached: {
-        assert(*pc() == Opcode::record_test_);
-        ObservedTest* feedback = (ObservedTest*)(pc() + 1);
-        feedback->seen = ObservedTest::Both;
+        auto& feedback = origin.function()->typeFeedback()->test(origin.idx());
+        feedback.seen = ObservedTest::Both;
         break;
     }
     case DeoptReason::Typecheck: {
-        assert(*pc() == Opcode::record_type_);
         if (val == symbol::UnknownDeoptTrigger)
             break;
-        ObservedValues* feedback = (ObservedValues*)(pc() + 1);
-        feedback->record(val);
-        if (TYPEOF(val) == PROMSXP) {
-            if (PRVALUE(val) == R_UnboundValue &&
-                feedback->stateBeforeLastForce < ObservedValues::promise)
-                feedback->stateBeforeLastForce = ObservedValues::promise;
-            else if (feedback->stateBeforeLastForce <
-                     ObservedValues::evaluatedPromise)
-                feedback->stateBeforeLastForce =
-                    ObservedValues::evaluatedPromise;
-        }
+        auto feedback = origin.function()->typeFeedback();
+
+        // FIXME: very similar code is in the recordTypeFeedbackImpl
```

Comment:
I would open a new issue to discuss this further and change it separately.
The function is also called from interp.cpp, where it differs a bit more
  ```
c->function()->typeFeedback()->record_type(idx, [&](auto& feedback) {
            if (feedback.stateBeforeLastForce < state) {
                feedback.stateBeforeLastForce = state;
            }
        });
```